# 第一部分：Python基础

操作系统：MacOS 26.1

Python版本：3.10.19

In [1]:
# 在 python 中使用 def 和 : 定义函数
def greet(name):
    return f"Hello, {name}!"
    # 或者 return "Hello" + ' ' + name + "!"

# 调用函数
message = greet("Wenting")
print(message)
# 输出: Hello, Wenting!

Hello, Wenting!


In [2]:
unicode_list = [36213, 29642, 31166]

print("你好，世界！我是", ''.join(chr(code) for code in unicode_list), "!")

你好，世界！我是 赵珊禾 !


## 1.1习题

In [3]:
def calculate_sales_tax(price, tax):
    cal_pri = (1 + tax / 100) * price
    return cal_pri

price = 170
tax = 12.5
cal_pri = calculate_sales_tax(price, tax)
print("价格为", price,"美元，税率为", tax,"%的商品最终售价为：", cal_pri, "美元")

价格为 170 美元，税率为 12.5 %的商品最终售价为： 191.25 美元


## 1.2习题

In [4]:
def calculate_bonus(level, salary):
    bonus_rate = [0, 5, 10, 15, 20]
    bonus = bonus_rate[level-1] * salary / 100
    return bonus

level = 3
salary = 28000
bonus = calculate_bonus(level, salary)
print("级绩效评级为", level,"级，工资为", salary,"美元的员工，奖金为：", bonus, "美元")

级绩效评级为 3 级，工资为 28000 美元的员工，奖金为： 2800.0 美元


# 第二部分：Excel线性回归

对region创建哑变量。

模型整体表现：

1. R^2值为0.751，说明模型能够解释75.1%的保险费用变异
2. 调整后R^2为0.749，表明模型拟合度良好
3. F统计量的p值为0.00，说明模型整体显著

各变量影响分析：

显著影响因素（p < 0.05）：

1. 年龄（age）：每增加1岁，保险费用增加约256.86美元（p=0.000）
2. BMI（bmi）：每增加1单位，保险费用增加约339.19美元（p=0.000）
3. 子女数量（children）：每多一个孩子，保险费用增加约475.50美元（p=0.001）
4. 吸烟状况（smoker）：吸烟者比非吸烟者保险费用高约23,850美元（p=0.000）
5. 东北地区（region_northeast）：比西南地区高约960.05美元（p=0.045）

不显著影响因素（p > 0.05）：

1. 性别（gender）：影响不显著（p=0.693）
2. 西北地区（region_northwest）：影响不显著（p=0.204）
3. 东南地区（region_southeast）：影响不显著（p=0.873）

主要发现：

1. 吸烟是影响保险费用的最重要因素
2. 年龄和BMI也是重要的正向影响因素
3. 地区差异中，只有东北地区与西南地区有显著差异
4. 性别对保险费用的影响在控制其他变量后不显著

关键发现总结：

最强影响因素：吸烟状况的影响系数最大（23,848.53），说明吸烟者的保险费用比非吸烟者平均高出约23,850美元，这是最显著的费用驱动因素。

重要正向影响因素：

1. BMI（339.19）：每增加1单位BMI，费用增加约339美元
2. 年龄（256.86）：每增加1岁，费用增加约257美元
3. 子女数量（475.50）：每多一个孩子，费用增加约476美元

地区差异：

1. 东北地区费用显著高于西南地区（+960.05美元）
2. 西北和东南地区与西南地区无显著差异

不显著因素：

1. 性别影响不显著（-131.31美元，p=0.693）
2. 东南地区费用略低于西南地区但不显著

业务建议：
1. 风险定价策略：重点关注吸烟者群体，可考虑提供戒烟激励计划来降低保险成本
2. 精准营销：针对高龄、高BMI人群设计专门的保险产品和服务
3. 区域策略：东北地区的保险费用基准可适当调高，反映地区差异
4. 产品设计：可考虑将子女数量作为家庭保险套餐的定价因素

该模型解释了75.1%的费用变异，具有较强的预测能力，可用于保险费用的预测和风险评估。

# 第三部分：Python线性回归

In [5]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

import statsmodels.api as sm

data = './insurance.csv'
insurance = pd.read_csv(data)

处理二分类变量

In [6]:
insurance['gender'] = insurance['gender'].map({'female': 0, 'male': 1})
insurance['smoker'] = insurance['smoker'].map({'no': 0, 'yes': 1})

对 region 创建哑变量

In [7]:
df = pd.get_dummies(insurance, columns=['region'], prefix='region', drop_first=False)

# print(df)

训练模型

In [8]:
y = df['charges']
X = df.drop(columns=['charges'])

# 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. 创建线性回归模型
model = LinearRegression()

# 4. 训练模型
model.fit(X_train, y_train)

# 5. 预测
y_pred = model.predict(X_test)

# 6. 模型评估
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("均方误差 (MSE):", mse)
print("R^2得分:", r2)
print("回归系数 (Coefficient):", model.coef_)
print("截距 (Intercept):", model.intercept_)

均方误差 (MSE): 33596915.85136147
R^2得分: 0.7835929767120723
回归系数 (Coefficient): [ 2.56975706e+02 -1.85916916e+01  3.37092552e+02  4.25278784e+02
  2.36511289e+04  4.59585244e+02  8.89079180e+01 -1.98279052e+02
 -3.50214110e+02]
截距 (Intercept): -12390.804294570275


输出模型摘要

In [9]:
X_train_numeric = X_train.astype(float)
y_train_numeric = y_train.astype(float)
model_ols = sm.OLS(y_train_numeric, X_train_numeric).fit()
print(model_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.740
Method:                 Least Squares   F-statistic:                     380.9
Date:                Sun, 09 Nov 2025   Prob (F-statistic):          1.32e-305
Time:                        01:28:51   Log-Likelihood:                -10845.
No. Observations:                1070   AIC:                         2.171e+04
Df Residuals:                    1061   BIC:                         2.175e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
age                256.9757     13.477  

根据上述代码输出，可得知：

结果与Excel中的结果近似一致，因为我将20%的数据划分为了测试集，因此结果不完全一致。

主要发现：

1. 吸烟是影响保险费用的最重要因素
2. 年龄和BMI也是重要的正向影响因素
3. 地区差异中，只有东北地区与西南地区有显著差异
4. 性别对保险费用的影响在控制其他变量后不显著

输出线性回归方程

In [10]:
equation_simple = f"charges = {model.intercept_:.2f}"
for i, col in enumerate(X.columns):
    if abs(model.coef_[i]) > 0.01:  # 只显示系数绝对值大于0.01的变量
        coef_sign = ' + ' if model.coef_[i] >= 0 else ' - '
        equation_simple += f"{coef_sign}{abs(model.coef_[i]):.2f}×{col}"

print("方程:", equation_simple)

方程: charges = -12390.80 + 256.98×age - 18.59×gender + 337.09×bmi + 425.28×children + 23651.13×smoker + 459.59×region_northeast + 88.91×region_northwest - 198.28×region_southeast - 350.21×region_southwest


预测的费用方程是Charges = 256.9757 X_age - 18.5917 X_gender + 337.0926 X_bmi  + 425.2788 X_children + 23651.1289 X_smoker + 459.5852 X_region_northeast + 88.9079 X_region_northwest - 198.2791 X_region_southeast - 350.2141 X_region_southwest - 12390.8043

In [11]:
# 创建测试数据
X_test2 = pd.DataFrame({
    'age': [32],
    'gender': [1],
    'bmi': [28.9],
    'children': [2],
    'smoker': [0],
    'region_northeast': [0],
    'region_northwest': [0],
    'region_southeast': [1],
    'region_southwest': [0]
})

# 确保列顺序与训练数据X一致
X_test2 = X_test2[X_train.columns]

# 进行预测
y_pred2 = model.predict(X_test2)
print("预测的保险费用:", y_pred2[0])

预测的保险费用: 6208.079866434184
